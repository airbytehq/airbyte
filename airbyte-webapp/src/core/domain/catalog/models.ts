export interface SyncSchemaField {
  cleanedName: string;
  type: string;
  key: string;
  path: string[];
  $ref?: string;
  airbyte_type?: string;
  format?: string;
  fields?: SyncSchemaField[];
}

export class SyncSchemaFieldObject {
  static isPrimitive(field: SyncSchemaField): boolean {
    return !(field.type === "object" || field.type === "array");
  }

  static isNestedField(field: SyncSchemaField): boolean {
    return field.path.length > 1;
  }
}
