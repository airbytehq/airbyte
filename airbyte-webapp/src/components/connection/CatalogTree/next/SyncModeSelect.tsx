import classNames from "classnames";
import { useMemo } from "react";
import { FormattedMessage } from "react-intl";

import { DropDownOptionDataItem } from "components/ui/DropDown";
import { PillSelect, PillButtonVariant } from "components/ui/PillSelect";

import { DestinationSyncMode, SyncMode } from "core/request/AirbyteClient";

import styles from "./SyncModeSelect.module.scss";

interface SyncModeValue {
  syncMode: SyncMode;
  destinationSyncMode: DestinationSyncMode;
}

export interface SyncModeOption {
  value: SyncModeValue;
}

interface SyncModeSelectProps {
  className?: string;
  onChange?: (option: DropDownOptionDataItem<SyncModeValue>) => void;
  options: SyncModeOption[];
  value: Partial<SyncModeValue>;
  variant?: PillButtonVariant;
}

export const SyncModeSelect: React.FC<SyncModeSelectProps> = ({ className, options, onChange, value, variant }) => {
  const pillSelectOptions = useMemo(() => {
    return options.map(({ value }) => {
      const { syncMode, destinationSyncMode } = value;
      return {
        label: [
          <FormattedMessage key={`syncMode.${syncMode}`} id={`syncMode.${syncMode}`} />,
          <FormattedMessage
            key={`destinationSyncMode.${destinationSyncMode}`}
            id={`destinationSyncMode.${destinationSyncMode}`}
          />,
        ],
        value,
      };
    });
  }, [options]);

  return (
    <PillSelect
      options={pillSelectOptions}
      value={value}
      onChange={onChange}
      className={classNames(styles.pillSelect, className)}
      variant={variant}
    />
  );
};
