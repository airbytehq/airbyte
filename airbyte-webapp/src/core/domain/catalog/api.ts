import { JSONSchema7 } from "json-schema";

import { AirbyteCatalog, AirbyteStreamAndConfiguration } from "../../request/AirbyteClient";

export type SchemaFields = JSONSchema7;

export interface SyncSchemaStream extends AirbyteStreamAndConfiguration {
  /**
   * This field is not returned from API and is used to track unique objects
   */
  id?: string;
}

export interface SyncSchema extends AirbyteCatalog {
  streams: SyncSchemaStream[];
}

export type Path = string[];
