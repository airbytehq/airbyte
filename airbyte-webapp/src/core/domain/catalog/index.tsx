export * from "./models";
export * from "./api";
export * from "./traverseSchemaToField";
