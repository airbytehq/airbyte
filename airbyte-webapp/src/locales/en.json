{
  "webapp.cannotReachServer": "Cannot reach server. The server may still be starting up.",

  "sidebar.sources": "Sources",
  "sidebar.destination": "Destination",
  "sidebar.help": "Help",
  "sidebar.docs": "Docs",

  "form.continue": "Continue",
  "form.emailOptional": "Your email (optional)",
  "form.email.placeholder": "you@company.com",
  "form.email.error": "This email address doesn’t seem correct.",
  "form.empty.error": "Empty field",
  "form.selectConnector": "Select a connector",
  "form.searchName": "search by name...",
  "form.noResult": "No result",
  "form.sourceName.placeholder": "Your source name",
  "form.destinationName.placeholder": "Your destination name",
  "form.connectionName.placeholder": "Name",
  "form.name": "Name",
  "form.sourceName.message": "Pick a name to help you identify this source in Airbyte",
  "form.destinationName.message": "Pick a name to help you identify this destination in Airbyte",
  "form.connectionName.message": "Pick a name to help you identify this source in Airbyte",
  "form.sourceType": "Source type",
  "form.destinationType": "Destination type",
  "form.connectionType": "Connector type",
  "form.frequency": "Sync frequency",
  "form.frequency.placeholder": "Select a frequency",
  "form.frequency.message": "Set how often Airbyte attempts to replicate data.",
  "form.dataSync": "Select the data you want to sync",
  "form.dataSync.message": "You’ll be able to change this later on.",
  "form.cancel": "Cancel",
  "form.delete": "Delete",
  "form.saveChanges": "Save changes",
  "form.discardChanges": "Discard changes",
  "form.every": "Every {value}",
  "form.testingConnection": "Testing connection...",
  "form.successTests": "All connection tests passed!",
  "form.failedTests": "The connection tests failed.",
  "form.host": "Host",
  "form.port": "Port",
  "form.user": "User",
  "form.password": "Password",
  "form.dbname": "DB Name",
  "form.postgres_host": "Host",
  "form.postgres_port": "Port",
  "form.postgres_username": "User",
  "form.postgres_password": "Password",
  "form.postgres_database": "DB Name",
  "form.postgres_schema": "Schema",
  "form.filter_dbs": "Filter DBs",
  "form.serviceAccount": "Service Account",
  "form.project_id": "Project ID",
  "form.dataset_id": "Dataset ID",
  "form.validate_records": "Validate Records",
  "form.credentials_json": "Credentials JSON",
  "form.disable_collection": "Disable Collection",
  "form.replication_method": "Replication Method",
  "form.validationError": "The received form did not pass validation",
  "form.someError": "Sorry. Something went wrong... ",
  "form.changesSaved": "Your changes were saved!",

  "preferences.title": "Specify your preferences",
  "preferences.anonymizeUsage": "Anonymize usage data collection",
  "preferences.collectData": "We collect data only for product improvements, see the <docs>docs</docs>.",
  "preferences.news": "News and feature updates",
  "preferences.security": "Security updates",
  "preferences.anonymizeData": "Anonymize my usage data.",
  "preferences.featureUpdates": "Receive feature updates.",
  "preferences.unsubcribeAnyTime": " You can unsubcribe any time.",
  "preferences.securityUpdates": "Receive emails about security updates.",

  "onboarding.title": "Let’s set up your first connection.",
  "onboarding.subtitle": "Connections are automated data pipelines that sync data from a source to a destination.",
  "onboarding.createSource": "Create a source",
  "onboarding.sourceSetUp": "Set up the source",
  "onboarding.sourceSetUp.buttonText": "Set up source",
  "onboarding.createDestination": "Create a destination",
  "onboarding.destinationSetUp": "Set up the destination",
  "onboarding.destinationSetUp.buttonText": "Set up destination",
  "onboarding.setUpConnection": "Set up connection",
  "onboarding.setConnection": "Set up the connection",
  "onboarding.instructionsLink": "See instructions how to connect {name}",
  "onboarding.fetchingSchema": "We are fetching the schema of your data source.\nThis should take less than a minute...",

  "sources.newSource": "+ New source",
  "sources.newSourceTitle": "New Source",
  "sources.selectSource": "Select a source",
  "sources.name": "Name",
  "sources.progress": "In progress",
  "sources.connector": "Connector",
  "sources.frequency": "Frequency",
  "sources.lastSync": "Last Sync",
  "sources.enabled": "Enabled",
  "sources.launch": "Launch",
  "sources.status": "Status",
  "sources.schema": "Schema",
  "sources.settings": "Settings",
  "sources.syncHistory": "Sync history",
  "sources.syncNow": "Sync now",
  "sources.source": "Source",
  "sources.disabled": "Disabled",
  "sources.noSync": "No sync yet",
  "sources.emptySchema": "Schema is empty",
  "sources.noSources": "Source list is empty",
  "sources.sourceSettings": "Source Settings",
  "sources.dataDelete": "No data will be deleted from your source.",
  "sources.deleteSource": "Delete this source",
  "sources.deleteConfirm": "Confirm source deletion",
  "sources.failed": "Failed",
  "sources.completed": "Completed",
  "sources.pending": "Pending",
  "sources.running": "Running",
  "sources.cancelled": "Cancelled",
  "sources.emptyLogs": "Empty data",
  "sources.hour": "{hour}h ",
  "sources.minute": "{minute}m ",
  "sources.second": "{second}s",
  "sources.deleteModalText": "This can not be un-done without a full re-sync. Note that:\n  - All past logs and configurations will be deleted\n  - Updates of new data will stop\n  - No existing data in the destination will be altered",

  "destination.destinationSettings": "Destination Settings"
}
