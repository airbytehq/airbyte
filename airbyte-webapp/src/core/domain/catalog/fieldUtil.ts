import { JSONSchema7Definition } from "json-schema";
import Status from "core/statuses";
import { CommonRequestError } from "core/request/CommonRequestError";

import { SourceDiscoverSchemaRead } from "./api";
import { SyncSchemaField } from "./models";
import { ConnectionNamespaceDefinition } from "../connection";
import { SOURCE_NAMESPACE_TAG } from "../connector/source";

function toInnerModel(
  result: SourceDiscoverSchemaRead
): SourceDiscoverSchemaRead {
  if (result.jobInfo?.status === Status.FAILED || !result.catalog) {
    // @ts-ignore address this case
    const e = new CommonRequestError(result);
    // Generate error with failed status and received logs
    e._status = 400;
    // @ts-ignore address this case
    e.response = result.jobInfo;
    throw e;
  }

  return result;
}

const traverseSchemaToField = (
  jsonSchema: JSONSchema7Definition,
  key: string
): SyncSchemaField[] => {
  // For the top level we should not insert an extra object
  return traverseJsonSchemaProperties(jsonSchema, key)[0].fields ?? [];
};

const traverseJsonSchemaProperties = (
  jsonSchema: JSONSchema7Definition,
  key: string,
  path: string[] = []
): SyncSchemaField[] => {
  if (typeof jsonSchema === "boolean") {
    return [];
  }

  let fields: SyncSchemaField[] | undefined;
  if (jsonSchema.properties) {
    fields = Object.entries(jsonSchema.properties)
      .flatMap(([k, schema]) =>
        traverseJsonSchemaProperties(schema, k, [...path, k])
      )
      .flat(2);
  }

  return [
    {
      cleanedName: key,
      path,
      key,
      fields,
      type:
        (Array.isArray(jsonSchema.type)
          ? jsonSchema.type.find((t) => t !== "null") ?? jsonSchema.type[0]
          : jsonSchema.type) ?? "null",
    },
  ];
};

type NamespaceOptions =
  | {
      namespaceDefinition:
        | ConnectionNamespaceDefinition.Source
        | ConnectionNamespaceDefinition.Destination;
      sourceNamespace?: string;
    }
  | {
      namespaceDefinition: ConnectionNamespaceDefinition.CustomFormat;
      namespaceFormat: string;
      sourceNamespace?: string;
    };

function getDestinationNamespace(opt: NamespaceOptions): string {
  const destinationSetting = "<destination schema>";
  switch (opt.namespaceDefinition) {
    case ConnectionNamespaceDefinition.Source:
      return opt.sourceNamespace ?? destinationSetting;
    case ConnectionNamespaceDefinition.Destination:
      return destinationSetting;
    case ConnectionNamespaceDefinition.CustomFormat:
      if (!opt.sourceNamespace?.trim()) {
        return destinationSetting;
      }

      return opt.namespaceFormat.replace(
        SOURCE_NAMESPACE_TAG,
        opt.sourceNamespace
      );
  }
}

export { getDestinationNamespace, traverseSchemaToField, toInnerModel };
