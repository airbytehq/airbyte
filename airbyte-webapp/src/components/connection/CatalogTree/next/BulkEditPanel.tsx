import classNames from "classnames";
import intersection from "lodash/intersection";
import React, { useMemo } from "react";
import { createPortal } from "react-dom";
import { FormattedMessage } from "react-intl";
import styled from "styled-components";

import { Header } from "components";
import { SUPPORTED_MODES } from "components/connection/ConnectionForm/formConfig";
import { Button } from "components/ui/Button";
import { Switch } from "components/ui/Switch";

import { SyncSchemaField, SyncSchemaFieldObject, SyncSchemaStream, traverseSchemaToField } from "core/domain/catalog";
import { DestinationSyncMode, SyncMode } from "core/request/AirbyteClient";
import { useBulkEditService } from "hooks/services/BulkEdit/BulkEditService";
import { useConnectionFormService } from "hooks/services/ConnectionForm/ConnectionFormService";

import styles from "./BulkEditPanel.module.scss";
import { StreamPathSelect } from "./StreamPathSelect";
import { SyncModeOption, SyncModeSelect } from "./SyncModeSelect";
import { pathDisplayName } from "../PathPopout";
import { HeaderCell } from "../styles";
import { flatten, getPathType } from "../utils";

interface SchemaHeaderProps {
  isActive: boolean;
}

const SchemaHeader = styled(Header)<SchemaHeaderProps>`
  position: fixed;
  bottom: ${(props) => (props.isActive ? 0 : "-100px")};
  left: 122px;
  z-index: 1000;
  width: calc(100% - 152px);
  height: unset;
  background: ${({ theme }) => theme.primaryColor};
  border-radius: 8px 8px 0 0;
  padding: 10px;
`;

export function calculateSharedFields(selectedBatchNodes: SyncSchemaStream[]) {
  const primitiveFieldsByStream = selectedBatchNodes.map(({ stream }) => {
    const traversedFields = traverseSchemaToField(stream?.jsonSchema, stream?.name);
    const flattenedFields = flatten(traversedFields);

    return flattenedFields.filter(SyncSchemaFieldObject.isPrimitive);
  });

  const pathMap = new Map<string, SyncSchemaField>();

  // calculate intersection of primitive fields across all selected streams
  primitiveFieldsByStream.forEach((fields, index) => {
    if (index === 0) {
      fields.forEach((field) => pathMap.set(pathDisplayName(field.path), field));
    } else {
      const fieldMap = new Set(fields.map((f) => pathDisplayName(f.path)));
      pathMap.forEach((_, k) => (!fieldMap.has(k) ? pathMap.delete(k) : null));
    }
  });

  return Array.from(pathMap.values());
}

export const getAvailableSyncModesOptions = (
  nodes: SyncSchemaStream[],
  syncModes?: DestinationSyncMode[]
): SyncModeOption[] =>
  SUPPORTED_MODES.filter(([syncMode, destinationSyncMode]) => {
    const supportableModes = intersection(nodes.flatMap((n) => n.stream?.supportedSyncModes));
    return supportableModes.includes(syncMode) && syncModes?.includes(destinationSyncMode);
  }).map(([syncMode, destinationSyncMode]) => ({
    value: { syncMode, destinationSyncMode },
  }));

export const BulkEditPanel: React.FC = () => {
  const {
    destDefinitionSpecification: { supportedDestinationSyncModes },
  } = useConnectionFormService();
  const { selectedBatchNodes, options, onChangeOption, onApply, isActive, onCancel } = useBulkEditService();
  const availableSyncModesOptions = useMemo(
    () => getAvailableSyncModesOptions(selectedBatchNodes, supportedDestinationSyncModes),
    [selectedBatchNodes, supportedDestinationSyncModes]
  );

  const primitiveFields: SyncSchemaField[] = useMemo(
    () => calculateSharedFields(selectedBatchNodes),
    [selectedBatchNodes]
  );

  const pkRequired = options.destinationSyncMode === DestinationSyncMode.append_dedup;
  const shouldDefinePk = selectedBatchNodes.every((n) => n.stream?.sourceDefinedPrimaryKey?.length === 0) && pkRequired;
  const cursorRequired = options.syncMode === SyncMode.incremental;
  const shouldDefineCursor = selectedBatchNodes.every((n) => !n.stream?.sourceDefinedCursor) && cursorRequired;
  const numStreamsSelected = selectedBatchNodes.length;

  const pkType = getPathType(pkRequired, shouldDefinePk);
  const cursorType = getPathType(cursorRequired, shouldDefineCursor);

  const paths = primitiveFields.map((f) => f.path);

  return createPortal(
    <SchemaHeader isActive={isActive}>
      <HeaderCell flex={0} className={classNames(styles.headerCell, styles.streamsCounterCell)}>
        <p className={classNames(styles.text, styles.streamsCountNumber)}>{numStreamsSelected}</p>
        <p className={classNames(styles.text, styles.streamsCountText)}>
          <FormattedMessage id="connection.streams" />
        </p>
      </HeaderCell>
      <HeaderCell flex={0} className={classNames(styles.headerCell, styles.syncCell)}>
        <p className={classNames(styles.text, styles.headerText)}>
          <FormattedMessage id="sources.sync" />
        </p>
        <div className={styles.syncCellContent}>
          <Switch
            variant="strong-blue"
            size="sm"
            checked={options.selected}
            onChange={() => onChangeOption({ selected: !options.selected })}
          />
        </div>
      </HeaderCell>
      <HeaderCell flex={1} className={styles.headerCell}>
        <p className={classNames(styles.text, styles.headerText)}>
          <FormattedMessage id="form.syncMode" />
        </p>
        <div className={styles.syncCellContent}>
          <SyncModeSelect
            className={styles.syncModeSelect}
            variant="strong-blue"
            value={{
              syncMode: options.syncMode,
              destinationSyncMode: options.destinationSyncMode,
            }}
            options={availableSyncModesOptions}
            onChange={({ value }) => onChangeOption({ ...value })}
          />
        </div>
      </HeaderCell>
      <HeaderCell flex={1} className={styles.headerCell}>
        <p className={classNames(styles.text, styles.headerText)}>
          <FormattedMessage id="form.cursorField" />
        </p>
        <div className={styles.syncCellContent}>
          <StreamPathSelect
            withSourceDefinedPill
            disabled={!cursorType}
            variant="strong-blue"
            isMulti={false}
            onPathChange={(path) => onChangeOption({ cursorField: path })}
            pathType={cursorType}
            paths={paths}
            path={options.cursorField}
          />
        </div>
      </HeaderCell>
      <HeaderCell flex={1} className={styles.headerCell}>
        <p className={classNames(styles.text, styles.headerText)}>
          <FormattedMessage id="form.primaryKey" />
        </p>
        <div className={styles.syncCellContent}>
          <StreamPathSelect
            withSourceDefinedPill
            disabled={!pkType}
            variant="strong-blue"
            isMulti
            onPathChange={(path) => onChangeOption({ primaryKey: path })}
            pathType={pkType}
            paths={paths}
            path={options.primaryKey}
          />
        </div>
      </HeaderCell>
      <HeaderCell flex={0} className={styles.buttonCell}>
        <Button className={styles.cancelButton} size="xs" variant="secondary" onClick={onCancel}>
          <FormattedMessage id="connectionForm.bulkEdit.cancel" />
        </Button>
        <Button className={styles.applyButton} size="xs" onClick={onApply}>
          <FormattedMessage id="connectionForm.bulkEdit.apply" />
        </Button>
      </HeaderCell>
    </SchemaHeader>,
    document.body
  );
};
