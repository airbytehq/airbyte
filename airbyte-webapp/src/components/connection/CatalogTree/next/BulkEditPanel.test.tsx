import { render } from "@testing-library/react";
import ReactDOM from "react-dom";
import { IntlProvider } from "react-intl";

import { SyncSchemaField, SyncSchemaStream } from "core/domain/catalog";
import en from "locales/en.json";

import { BulkEditPanel, calculateSharedFields, getAvailableSyncModesOptions } from "./BulkEditPanel";
import { SyncModeOption } from "./SyncModeSelect";

const MOCK_NODE_1: SyncSchemaStream = {
  stream: {
    name: "balance_transactions",
    jsonSchema: {
      type: ["null", "object"],
      properties: {
        id: {
          type: ["null", "string"],
        },
        fee: {
          type: ["null", "integer"],
        },
        net: {
          type: ["null", "integer"],
        },
        type: {
          type: ["null", "string"],
        },
        amount: {
          type: ["null", "integer"],
        },
        object: {
          type: ["null", "string"],
        },
        source: {
          type: ["null", "string"],
        },
        status: {
          type: ["null", "string"],
        },
        created: {
          type: ["null", "integer"],
        },
        currency: {
          type: ["null", "string"],
        },
        description: {
          type: ["null", "string"],
        },
        fee_details: {
          type: ["null", "array"],
          items: {
            type: ["null", "object"],
            properties: {
              type: {
                type: ["null", "string"],
              },
              amount: {
                type: ["null", "integer"],
              },
              currency: {
                type: ["null", "string"],
              },
              application: {
                type: ["null", "string"],
              },
              description: {
                type: ["null", "string"],
              },
            },
          },
        },
        available_on: {
          type: ["null", "integer"],
        },
        exchange_rate: {
          type: ["null", "number"],
        },
        sourced_transfers: {
          type: ["null", "array"],
          items: {},
        },
      },
    },
    supportedSyncModes: ["full_refresh", "incremental"],
    sourceDefinedCursor: true,
    defaultCursorField: ["created"],
    sourceDefinedPrimaryKey: [["id"]],
  },
  config: {
    syncMode: "incremental",
    cursorField: ["created"],
    destinationSyncMode: "append_dedup",
    primaryKey: [["id"]],
    aliasName: "balance_transactions",
    selected: true,
  },
  id: "0",
};
const MOCK_NODE_2: SyncSchemaStream = {
  stream: {
    name: "bank_accounts",
    jsonSchema: {
      type: ["null", "object"],
      properties: {
        id: {
          type: ["null", "string"],
        },
        last4: {
          type: ["null", "string"],
        },
        object: {
          type: ["null", "string"],
        },
        status: {
          type: ["null", "string"],
        },
        country: {
          type: ["null", "string"],
        },
        currency: {
          type: ["null", "string"],
        },
        customer: {
          type: ["null", "string"],
        },
        metadata: {
          type: ["null", "object"],
          properties: {},
        },
        bank_name: {
          type: ["null", "string"],
        },
        fingerprint: {
          type: ["null", "string"],
        },
        routing_number: {
          type: ["null", "string"],
        },
        account_holder_name: {
          type: ["null", "string"],
        },
        account_holder_type: {
          type: ["null", "string"],
        },
      },
    },
    supportedSyncModes: ["full_refresh"],
    defaultCursorField: [],
    sourceDefinedPrimaryKey: [["id"]],
  },
  config: {
    syncMode: "full_refresh",
    cursorField: [],
    destinationSyncMode: "overwrite",
    primaryKey: [["id"]],
    aliasName: "bank_accounts",
    selected: true,
  },
  id: "1",
};

jest.mock("hooks/services/BulkEdit/BulkEditService", () => ({
  useBulkEditService: () => ({
    selectedBatchNodes: [MOCK_NODE_1, MOCK_NODE_2],
    options: {
      selected: false,
    },
    isActive: true,
    onChangeOption: jest.fn(),
    onApply: jest.fn(),
    onCancel: jest.fn(),
  }),
}));

jest.mock("hooks/services/ConnectionForm/ConnectionFormService", () => ({
  useConnectionFormService: () => ({ destDefinitionSpecification: ["overwrite", "append", "append_dedup"] }),
}));

const renderBulkEditPanel = () =>
  render(
    <IntlProvider locale="en" messages={en}>
      <BulkEditPanel />
    </IntlProvider>
  );

describe("<BulkEditPanel />", () => {
  beforeAll(() => {
    // @ts-expect-error Okay for test
    ReactDOM.createPortal = (element) => {
      return element;
    };
  });

  it("should render", () => {
    const component = renderBulkEditPanel();
    expect(component).toMatchSnapshot();
  });

  it("calculateSharedFields should work correctly", () => {
    const expectedResult: SyncSchemaField[] = [
      { cleanedName: "id", path: ["id"], key: "id", type: "string" },
      { cleanedName: "object", path: ["object"], key: "object", type: "string" },
      { cleanedName: "status", path: ["status"], key: "status", type: "string" },
      { cleanedName: "currency", path: ["currency"], key: "currency", type: "string" },
    ];
    const actualResult = calculateSharedFields([MOCK_NODE_1, MOCK_NODE_2]);
    expect(actualResult).toEqual(expectedResult);
  });

  it("getAvailableSyncModesOptions should work correctly", () => {
    const expectedResult: SyncModeOption[] = [
      { value: { syncMode: "incremental", destinationSyncMode: "append_dedup" } },
      { value: { syncMode: "full_refresh", destinationSyncMode: "overwrite" } },
      { value: { syncMode: "incremental", destinationSyncMode: "append" } },
      { value: { syncMode: "full_refresh", destinationSyncMode: "append" } },
    ];
    const actualResult = getAvailableSyncModesOptions(
      [MOCK_NODE_1, MOCK_NODE_2],
      ["overwrite", "append", "append_dedup"]
    );
    expect(actualResult).toEqual(expectedResult);
  });
});
