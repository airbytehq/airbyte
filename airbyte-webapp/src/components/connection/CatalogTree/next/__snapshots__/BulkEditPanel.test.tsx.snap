// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BulkEditPanel /> should render 1`] = `
<body>
  <div>
    <div
      class="<removed-for-snapshot-test>"
    >
      <div
        class="<removed-for-snapshot-test>"
      >
        <p
          class="<removed-for-snapshot-test>"
        >
          2
        </p>
        <p
          class="<removed-for-snapshot-test>"
        >
          Streams
        </p>
      </div>
      <div
        class="<removed-for-snapshot-test>"
      >
        <p
          class="<removed-for-snapshot-test>"
        >
          Sync
        </p>
        <div
          class="<removed-for-snapshot-test>"
        >
          <label
            class="<removed-for-snapshot-test>"
          >
            <input
              aria-checked="false"
              class="<removed-for-snapshot-test>"
              type="checkbox"
              value=""
            />
            <span
              class="<removed-for-snapshot-test>"
            />
          </label>
        </div>
      </div>
      <div
        class="<removed-for-snapshot-test>"
      >
        <p
          class="<removed-for-snapshot-test>"
        >
          Sync mode
        </p>
        <div
          class="<removed-for-snapshot-test>"
        >
          <div
            class="<removed-for-snapshot-test>"
            role="combobox"
          >
            <span
              class="<removed-for-snapshot-test>"
              id="react-select-2-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="<removed-for-snapshot-test>"
            />
            <div>
              <span
                class="<removed-for-snapshot-test>"
              >
                <button
                  class="<removed-for-snapshot-test>"
                  type="button"
                >
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    <span
                      class="<removed-for-snapshot-test>"
                      data-type="text"
                    />
                  </div>
                  <svg
                    aria-hidden="true"
                    class="<removed-for-snapshot-test>"
                    data-icon="caret-down"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 320 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M310.6 246.6l-127.1 128C176.4 380.9 168.2 384 160 384s-16.38-3.125-22.63-9.375l-127.1-128C.2244 237.5-2.516 223.7 2.438 211.8S19.07 192 32 192h255.1c12.94 0 24.62 7.781 29.58 19.75S319.8 237.5 310.6 246.6z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="<removed-for-snapshot-test>"
      >
        <p
          class="<removed-for-snapshot-test>"
        >
          Cursor field
        </p>
        <div
          class="<removed-for-snapshot-test>"
        >
          <div
            class="<removed-for-snapshot-test>"
            role="combobox"
          >
            <span
              class="<removed-for-snapshot-test>"
              id="react-select-3-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="<removed-for-snapshot-test>"
            />
            <div>
              <span
                class="<removed-for-snapshot-test>"
              >
                <button
                  class="<removed-for-snapshot-test>"
                  disabled=""
                  type="button"
                >
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    <span
                      class="<removed-for-snapshot-test>"
                      data-type="text"
                    >
                      Not available
                    </span>
                  </div>
                  <svg
                    aria-hidden="true"
                    class="<removed-for-snapshot-test>"
                    data-icon="caret-down"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 320 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M310.6 246.6l-127.1 128C176.4 380.9 168.2 384 160 384s-16.38-3.125-22.63-9.375l-127.1-128C.2244 237.5-2.516 223.7 2.438 211.8S19.07 192 32 192h255.1c12.94 0 24.62 7.781 29.58 19.75S319.8 237.5 310.6 246.6z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="<removed-for-snapshot-test>"
      >
        <p
          class="<removed-for-snapshot-test>"
        >
          Primary key
        </p>
        <div
          class="<removed-for-snapshot-test>"
        >
          <div
            class="<removed-for-snapshot-test>"
            role="combobox"
          >
            <span
              class="<removed-for-snapshot-test>"
              id="react-select-4-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="<removed-for-snapshot-test>"
            />
            <div>
              <span
                class="<removed-for-snapshot-test>"
              >
                <button
                  class="<removed-for-snapshot-test>"
                  disabled=""
                  type="button"
                >
                  <div
                    class="<removed-for-snapshot-test>"
                  >
                    <span
                      class="<removed-for-snapshot-test>"
                      data-type="text"
                    >
                      Not available
                    </span>
                  </div>
                  <svg
                    aria-hidden="true"
                    class="<removed-for-snapshot-test>"
                    data-icon="caret-down"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 320 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M310.6 246.6l-127.1 128C176.4 380.9 168.2 384 160 384s-16.38-3.125-22.63-9.375l-127.1-128C.2244 237.5-2.516 223.7 2.438 211.8S19.07 192 32 192h255.1c12.94 0 24.62 7.781 29.58 19.75S319.8 237.5 310.6 246.6z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="<removed-for-snapshot-test>"
      >
        <button
          class="<removed-for-snapshot-test>"
        >
          <div
            class="<removed-for-snapshot-test>"
          >
            Cancel
          </div>
        </button>
        <button
          class="<removed-for-snapshot-test>"
        >
          <div
            class="<removed-for-snapshot-test>"
          >
            Apply
          </div>
        </button>
      </div>
    </div>
  </div>
</body>
`;
